### DISCUSSION TITLE
Abstract `owner` condition for use alongside NestedElementManager

### DISCUSSION THREAD

**Discussion by ryansupercool** [craftcms/cms#15882] [Ideas] [answered]
The `owner` condition used in `beforePrepare` is a bit complicated and is duplicated (with minor variants to the names of the tables) in various element queries in the Craft org.

We're looking at implementing `NestedElementManager` and the associated `owner` method in our own Elements and was wondering if this part of the query could be abstracted on Craft's end to save us (and others) having to implement it (copy and paste) into our own code?

https://github.com/craftcms/cms/blob/68557cb0dbe84beb628274e58d1a16442413ad97/src/elements/db/EntryQuery.php#L1188-L1232

**Comment by ryansupercool** [comment]
Along the same line could `NestedElementManager` or a helper be provided to help setting ownership in an elements `afterSave` method? It feels a little unsafe for a plugin to be writing directly to the element owners table. Could this be handle via a service method?

https://github.com/craftcms/cms/blob/68557cb0dbe84beb628274e58d1a16442413ad97/src/elements/Address.php#L672-L724

https://github.com/craftcms/commerce/blob/a156731711ab80f9197a14f8832420908b4a9f4f/src/elements/Variant.php#L937-L989

https://github.com/craftcms/stripe/blob/71bcbee30203f432825e24de9f42702f163c4b99/src/elements/Price.php#L469-L519

**Comment by brandonkelly** [comment, marked as answer]
Thanks for the suggestions! Just added `NestedElementQueryTrait` and `NestedElementTrait::saveOwnership()` for Craft 5.5 (#15894).

**Reply by rynpsc** [reply]
Amazing. Thanks!

Would this also work for element queries that don't have a `fieldId` such as in the Stripe and Commerce plugins?

https://github.com/craftcms/commerce/blob/5.x/src/elements/db/VariantQuery.php#L416
https://github.com/craftcms/stripe/blob/1.x/src/elements/db/PriceQuery.php#L500

**Reply by brandonkelly** [reply]
Right now it’s specifically built for nested elements that are managed via custom fields, at least optionally. (Addresses go both ways.)

**Reply by ryansupercool** [reply]
I've just been looking at this again and the elements we are dealing with work very much like Products/Variants in Commerce and Products/Prices in Stripe. Both of those plugins are also managing owership with almost an identical chunk of code to the now abstracted method with exception of the `fieldId`. Other plugins appear to be duplicating this bits of this code as well such as Verbb Events.

Do you think there's scope for Craft core abstracting the save owership logic for elements that aren't being used in the context of a field?

https://github.com/craftcms/stripe/blob/f7440e5d920f2daabd445ca6dcb9cf3176c14cac/src/elements/Price.php#L470-L520
https://github.com/craftcms/commerce/blob/272475a8b385896e21282e691ec764387b03b191/src/elements/Variant.php#L937-L989

https://github.com/craftcms/stripe/blob/f7440e5d920f2daabd445ca6dcb9cf3176c14cac/src/elements/db/PriceQuery.php#L546-L589
